src/soroban/assembled_transaction.ts

```diff
@@ -0,0 +1,717 @@
+  authorizeEntry,
+  hash,
+  xdr,
+} from "..";
+import type { Memo, MemoType, Transaction } from "..";
+
+type Tx = Transaction<Memo<MemoType>, Operation[]>;
+
+type SendTx = SorobanRpc.Api.SendTransactionResponse;
+type GetTx = SorobanRpc.Api.GetTransactionResponse;
+
+/**
+ * Error interface containing the error message
+ */
+interface ErrorMessage {
+  message: string;
+}
+
+interface Result<T, E extends ErrorMessage> {
+  unwrap(): T;
+  unwrapErr(): E;
+  isOk(): boolean;
+  isErr(): boolean;
+}
+class Ok<T> implements Result<T, never> {
+  constructor(readonly value: T) {}
+  unwrapErr(): never { throw new Error("No error") }
+  unwrap() { return this.value }
+  isOk() { return true }
+  isErr() { return false }
+}
+class Err<E extends ErrorMessage> implements Result<never, E> {
+  constructor(readonly error: E) {}
+  unwrapErr() { return this.error }
+  unwrap(): never { throw new Error(this.error.message) }
+  isOk() { return false }
+  isErr() { return true }
+}
+
+export type MethodOptions = {
+  /**
+   * The fee to pay for the transaction. Default: BASE_FEE
+   */
+  fee?: number;
+};
+
+export type AssembledTransactionOptions<T = string> = MethodOptions &
+  ContractClientOptions & {
+    method: string;
+    args?: any[];
+    parseResultXdr: (xdr: xdr.ScVal) => T;
+  };
+
+export const NULL_ACCOUNT =
+  "GAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAWHF";
+
+export class AssembledTransaction<T> {
+  public raw?: Tx;
+  private simulation?: SorobanRpc.Api.SimulateTransactionResponse;
+  private simulationResult?: SorobanRpc.Api.SimulateHostFunctionResult;
+  private simulationTransactionData?: xdr.SorobanTransactionData;
+  private server: SorobanRpc.Server;
+
+  static ExpiredStateError = class ExpiredStateError extends Error {}
+  static NeedsMoreSignaturesError = class NeedsMoreSignaturesError extends Error {}
+  static WalletDisconnectedError = class WalletDisconnectedError extends Error {}
+  static SendResultOnlyError = class SendResultOnlyError extends Error {}
+  static SendFailedError = class SendFailedError extends Error {}
+  static NoUnsignedNonInvokerAuthEntriesError = class NoUnsignedNonInvokerAuthEntriesError extends Error {}
+
+  static Result = { Ok, Err }
+
+  toJSON() {
+    return JSON.stringify({
+      method: this.options.method,
+      tx: this.raw?.toXDR(),
+      simulationResult: {
+        auth: this.simulationData.result.auth.map((a) => a.toXDR("base64")),
+        retval: this.simulationData.result.retval.toXDR("base64"),
+      },
+      simulationTransactionData:
+        this.simulationData.transactionData.toXDR("base64"),
+    });
+  }
+
+  static fromJSON<T>(
+    options: Omit<AssembledTransactionOptions<T>, "args">,
+    {
+      tx,
+      simulationResult,
+      simulationTransactionData,
+    }: {
+      tx: XDR_BASE64;
+      simulationResult: {
+        auth: XDR_BASE64[];
+        retval: XDR_BASE64;
+      };
+      simulationTransactionData: XDR_BASE64;
+    }
```

Comment:
Could this just be `Transaction` and `Api.SimulateTransactionSuccessResponse`?

Reply:
Maybe. [Again](https://github.com/stellar/js-stellar-sdk/pull/891#discussion_r1434521447), this was working around errors we were getting while serializing/deserializing.